File: lab_set.cpp
#include <exception>
#include <iostream>
#include <string>
using namespace std;

class Set
{
private:
	const int maxSize;
	char *set;
	int items;

public:
	class Full: public exception
	{
	public:
		virtual const char* what() const throw()
  		{
    		return "Set is full";
  		}
	};

	/** Initialise the Set, maxSize is the maximum number of
		values that can be stored in the Set at any time **/
	Set( const int maxSize ) : maxSize(maxSize)
	{
		items = 0;
		set = new char[ maxSize ];  // create the array to store the set
		for( int i=0; i<maxSize; ++i )
		{
			set[i] = 0;					// set the starting value of every element to 0
		}
	}	

	~Set()
	{
		delete [] set;
	}

	/** Checks to see if value is in the set or not,
			returns True if it is and False if it isn't **/
	bool is_in( char value )
	{
		// COMPLETE ME

		return false;
	}

	/** Add value to the set, raises set.Full 
			exception if set is full **/
	void add( char value )
	{
		// COMPLETE ME
	}

	/** Returns the number of values currently stored in the 
		set **/
	int num_items()
	{
		return items;
	}

	/** Removes value from the set if it is present,
		returns True if value did exist and False if it did not **/
	bool remove( char value )
	{
		// COMPLETE ME

		return false;
	}
};

int main()
{
	Set s( 5 );
	
	// === adding test ======
	cout << "Adding a" << endl;
	s.add('a');
	if( s.num_items() != 1 )
	{
		cerr << "Error - set should have 1 item but claims it has " << s.num_items() << endl;
	}

	if( !s.is_in('a') )
	{
		cerr << "Error - set should contain 'a' but claims it doesn't" << endl;
	}

	cout << "Adding b" << endl;
	s.add('b');
	if( s.num_items() != 2 )
	{
		cerr << "Error - set should have 2 item but claims it has " << s.num_items() << endl;
	}

	if( !s.is_in('b') )
	{
		cerr << "Error - set should contain 'b' but claims it doesn't" << endl;
	}

	// === duplicate test ======
	cout << "Adding a again" << endl;
	s.add('a');
	if( s.num_items() != 2 )
	{
		cerr << "Error - set should have 2 item but claims it has " << s.num_items() << endl;
	}

	if( !s.is_in('a') )
	{
		cerr << "Error - set should contain 'a' but claims it doesn't" << endl;
	}

	if( s.is_in('c') )
	{
		cerr << "Error - set doesn't contain 'c' but claims it does" << endl;
	}

	if( s.remove('c') )
	{
		cerr << "Error - set doesn't contain 'c' but claimed that remove() just got rid of it" << endl;
	}

	// === removing test ======
	cout << "Removing a" << endl;
	if( !s.remove('a') )
	{
		cerr << "Error - set contained 'a' but set claimed that remove() can't get rid of it" << endl;
	}

	if( s.num_items() != 1 )
	{
		cerr << "Error - set should have 1 item but claims it has " << s.num_items() << endl;
	}

	return 0;
}

File: lab_set.py
class Set:
	class Full(Exception):
		pass

	def __init__( self, maxSize ):
		""" Initialise the set, maxSize is the maximum number of
			values that can be stored in the set at any time """
		self.__maxSize = maxSize
		self.__set = [None]*self.__maxSize
		self.__items = 0

	def is_in( self, value ):
		""" Checks to see if value is in the set or not,
			returns True if it is and False if it isn't """
		# COMPLETE ME

		return False

	def add( self, value ):
		""" Add value to the set, raises set.Full 
			exception if set is full """
		# COMPLETE ME

	def num_items( self ):
		""" Returns the number of values currently stored in the 
			set """
		return self.__items

	def remove( self, value ):
		""" Removes value from the set if it is present,
			returns True if value did exist and False if it did not """
		
		# COMPLETE ME
		return False



if __name__ == '__main__':

	s = Set( 5 )
	
	# === adding test ======
	print( 'Adding a' )
	s.add('a')
	if s.num_items() != 1:
		print( 'Error - set should have 1 item but claims it has %d' % s.num_items() )

	if not s.is_in('a'):
		print( 'Error - set should contain \'a\' but claims it doesn\'t' )

	print( 'Adding b' )
	s.add('b')
	if s.num_items() != 2:
		print( 'Error - set should have 2 item but claims it has %d' % s.num_items() )

	if not s.is_in('b'):
		print( 'Error - set should contain \'b\' but claims it doesn\'t' )

	print( 'Adding a again' )
	s.add('a')
	if s.num_items() != 2:
		print( 'Error - set should have 2 item but claims it has %d' % s.num_items() )

	if not s.is_in('a'):
		print( 'Error - set should contain \'a\' but claims it doesn\'t' )

	if s.is_in('c'):
		print( 'Error - set doesn\'t contain \'c\' but claims it does' )

	if s.remove('c'):
		print( 'Error - set doesn\'t contain \'c\' but claimed that remove() just got rid of it' )


	print( 'Removing a ')
	if not s.remove( 'a' ):
		print( 'Error - set contained \'a\' but set claimed that remove() can\'t get rid of it' )

	if s.num_items() != 1:
		print( 'Error - set should have 1 item but claims it has %d' % s.num_items() )

